File: 31-next-permutation/next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        ind = n-2

        while (ind >= 0 and nums[ind]>=nums[ind+1]):
            ind-=1
        
        if ind >= 0:
            j = n-1
            for i in range(j,-1,-1):
                if nums[i] > nums[ind]:
                    break
            
            nums[ind],nums[i] = nums[i],nums[ind]
        
        nums[ind+1:] = nums[ind+1:][::-1]
        return nums

File: 31-next-permutation/next-permutation.java
class Solution {
    public void reverse(int[] arr , int start ,int end){
        while(start < end){
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;

            start++;
            end--;
        }
    }
    public void nextPermutation(int[] nums) {

        // find the breaking point
        int n = nums.length;
        int i = n-2;

        while(i>=0 && nums[i]>=nums[i+1]){
            i--;
        }

        if(i>=0){
            int j = n-1;
            // find the slight grater then ith element

            while(nums[j] <= nums[i]){
                j--;
            }
            // swap i -- j
            int temp = nums[i];
            nums[i] =nums[j];
            nums[j]=temp;
        }

        reverse(nums,i+1,n-1);



    }
}